cstrml/staking/src/lib.rs

```diff
@@ -2437,12 +2466,37 @@ impl<T: Config> historical::SessionManager<T::AccountId, Exposure<T::AccountId,
 
 impl<T: Config> swork::Works<T::AccountId> for Module<T> {
     fn report_works(workload_map: BTreeMap<T::AccountId, u128>, total_workload: u128) {
-        for (v_stash, _) in <Validators<T>>::iter() {
-            let v_own_workload = workload_map.get(&v_stash).unwrap_or(&0u128);
-            Self::upsert_stake_limit(
-                &v_stash,
-                Self::stake_limit_of(*v_own_workload, total_workload),
-            );
+        let total_issuance = T::Currency::total_issuance();
+        let stake_limit_ratio = Self::stake_limit_ratio();
+        // If effective staking ratio is smaller than some value, we should increase the total stake limit
+        let extra_stake_limit_ratio = Self::add_extra_ratio_according_to_effective_staking(total_issuance.clone());
+        let total_stake_limit = TryInto::<u128>::try_into((stake_limit_ratio * total_issuance).saturating_add(extra_stake_limit_ratio * total_issuance))
+            .ok()
+            .unwrap();
+        // Judge if we're in stage one or stage two
+        let is_stage_one = total_workload.saturating_mul(T::SPowerRatio::get()) < total_stake_limit;
```

Comment:
I don't this this judging should be appeared in here, abstract this juding into a function